service/history/historyEngine.go

```diff
@@ -294,6 +302,56 @@ func (e *historyEngineImpl) GetWorkflowExecutionNextEventID(
 		RunId:      request.Execution.RunId,
 	}
 
+	response, err := e.getWorkflowExecutionNextEventID(domainID, execution)
+	if err != nil {
+		return nil, err
+	}
+
+	// expectedNextEventID is 0 when caller want to get the current next event ID without blocking
+	expectedNextEventID := request.GetExpectedNextEventId()
+
+	// if caller decide to long poll on workflow execution
+	// and the event ID we are looking for is smaller than current next event ID
+	if expectedNextEventID >= response.GetEventId() && response.GetIsWorkflowRunning() {
+		subscriberID, channel, err := e.historyEventNotifier.WatchHistoryEvent(&domainID, &execution)
+		if err != nil {
+			return nil, err
+		}
+		defer e.historyEventNotifier.UnwatchHistoryEvent(&domainID, &execution, subscriberID)
+
+		// check again in case the next event ID is updated
+		response, err = e.getWorkflowExecutionNextEventID(domainID, execution)
```

Comment:
What about calling WatchHistoryEvent and then getWorkflowExecutionNextEventID, this is going to eliminate race condition without requiring a second call.

Reply:
yes, that will eliminate the one call to getWorkflowExecutionNextEventID

my concern is, most of the time, the expected next event ID in the request will be met, meaning no long poll is necessary, that is the main reason for me doing a "Double-checked locking"